From 0cd6da883cc0face570c4738e7ae918f2010c177 Mon Sep 17 00:00:00 2001
From: Nell <xianbin.zhu@spacemit.com>
Date: Wed, 3 Apr 2024 19:39:15 +0800
Subject: [PATCH 0547/1200] k1:cpufreq-cooling: refine some code for cpu
 cooling

1. we'd better acquire the lock throughout the entire CPU frequency modulation process
2. we started implementing temperature control strategies at 95 degrees

Change-Id: I402b438b928991a2df8dd1678ba83bca983ae005
---
 arch/riscv/boot/dts/spacemit/k1-x.dtsi        | 80 +++++--------------
 arch/riscv/boot/dts/spacemit/k1-x_MINI-PC.dts |  4 +-
 arch/riscv/boot/dts/spacemit/k1-x_deb1.dts    |  4 +-
 arch/riscv/boot/dts/spacemit/k1-x_deb2.dts    |  4 +-
 arch/riscv/boot/dts/spacemit/k1-x_evb.dts     |  4 +-
 arch/riscv/boot/dts/spacemit/k1-x_hs450.dts   |  4 +-
 arch/riscv/boot/dts/spacemit/k1-x_kx312.dts   |  4 +-
 arch/riscv/boot/dts/spacemit/k1-x_mingo.dts   |  4 +-
 drivers/cpufreq/spacemit-cpufreq.c            | 14 ++--
 9 files changed, 42 insertions(+), 80 deletions(-)

diff --git a/arch/riscv/boot/dts/spacemit/k1-x.dtsi b/arch/riscv/boot/dts/spacemit/k1-x.dtsi
index e2ad8556c0c0..e65dd2188ac8 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x.dtsi
+++ b/arch/riscv/boot/dts/spacemit/k1-x.dtsi
@@ -2033,30 +2033,18 @@ cluster0_thermal {
 
 				trips {
 					cls0_trip0: cls0-trip-point0 {
-						temperature = <55000>;
+						temperature = <95000>;
 						hysteresis = <5000>;
 						type = "passive";
 					};
 
 					cls0_trip1: cls0-trip-point1 {
-						temperature = <70000>;
-						hysteresis = <5000>;
-						type = "passive";
-					};
-
-					cls0_trip2: cls0-trip-point2 {
-						temperature = <85000>;
-						hysteresis = <5000>;
-						type = "passive";
-					};
-
-					cls0_trip3: cls0-trip-point3 {
 						temperature = <105000>;
 						hysteresis = <5000>;
 						type = "passive";
 					};
 
-					cls0_trip4: cls0-trip-point4 {
+					cls0_trip2: cls0-trip-point2 {
 						temperature = <115000>;
 						hysteresis = <5000>;
 						type = "critical";
@@ -2064,9 +2052,11 @@ cls0_trip4: cls0-trip-point4 {
 				};
 
 				cooling-maps {
-/*
 					map0 {
-						trip = <&cls0_trip1>;
+						trip = <&cls0_trip0>;
+/*
+						cooling-device = <&{/cpus/cpu@3/thermal-hotplug} 0 1>;
+*/
 						cooling-device = <&cpu_0 1 2>,
 								 <&cpu_1 1 2>,
 								 <&cpu_2 1 2>,
@@ -2074,22 +2064,14 @@ map0 {
 					};
 
 					map1 {
-						trip = <&cls0_trip2>;
-						cooling-device = <&cpu_0 3 4>,
-								 <&cpu_1 3 4>,
-								 <&cpu_2 3 4>,
-								 <&cpu_3 3 4>;
-					};
-*/
-					map2 {
-						trip = <&cls0_trip3>;
+						trip = <&cls0_trip1>;
 /*
 						cooling-device = <&{/cpus/cpu@3/thermal-hotplug} 0 1>;
 */
-						cooling-device = <&cpu_0 1 1>,
-								 <&cpu_1 1 1>,
-								 <&cpu_2 1 1>,
-								 <&cpu_3 1 1>;
+						cooling-device = <&cpu_0 3 3>,
+								 <&cpu_1 3 3>,
+								 <&cpu_2 3 3>,
+								 <&cpu_3 3 3>;
 					};
 				};
 			};
@@ -2101,30 +2083,18 @@ cluster1_thermal {
 
 				trips {
 					cls1_trip0: cls1-trip-point0 {
-						temperature = <55000>;
+						temperature = <95000>;
 						hysteresis = <5000>;
 						type = "passive";
 					};
 
 					cls1_trip1: cls1-trip-point1 {
-						temperature = <70000>;
-						hysteresis = <5000>;
-						type = "passive";
-					};
-
-					cls1_trip2: cls1-trip-point2 {
-						temperature = <85000>;
-						hysteresis = <5000>;
-						type = "passive";
-					};
-
-					cls1_trip3: cls1-trip-point3 {
 						temperature = <105000>;
 						hysteresis = <5000>;
 						type = "passive";
 					};
 
-					cls1_trip4: cls1-trip-point4 {
+					cls1_trip2: cls1-trip-point2 {
 						temperature = <115000>;
 						hysteresis = <5000>;
 						type = "critical";
@@ -2132,9 +2102,11 @@ cls1_trip4: cls1-trip-point4 {
 				};
 
 				cooling-maps {
-/*
 					map0 {
-						trip = <&cls1_trip1>;
+						trip = <&cls1_trip0>;
+/*
+						cooling-device = <&{/cpus/cpu@7/thermal-hotplug} 0 1>;
+*/
 						cooling-device = <&cpu_4 1 2>,
 								 <&cpu_5 1 2>,
 								 <&cpu_6 1 2>,
@@ -2142,22 +2114,14 @@ map0 {
 					};
 
 					map1 {
-						trip = <&cls1_trip2>;
-						cooling-device = <&cpu_4 3 4>,
-								 <&cpu_5 3 4>,
-								 <&cpu_6 3 4>,
-								 <&cpu_7 3 4>;
-					};
-*/
-					map2 {
-						trip = <&cls1_trip3>;
+						trip = <&cls1_trip1>;
 /*
 						cooling-device = <&{/cpus/cpu@7/thermal-hotplug} 0 1>;
 */
-						cooling-device = <&cpu_4 1 1>,
-								 <&cpu_5 1 1>,
-								 <&cpu_6 1 1>,
-								 <&cpu_7 1 1>;
+						cooling-device = <&cpu_4 3 3>,
+								 <&cpu_5 3 3>,
+								 <&cpu_6 3 3>,
+								 <&cpu_7 3 3>;
 					};
 				};
 			};
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_MINI-PC.dts b/arch/riscv/boot/dts/spacemit/k1-x_MINI-PC.dts
index 106be807779c..fb00a67666ea 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_MINI-PC.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_MINI-PC.dts
@@ -681,11 +681,11 @@ &rcpu {
 	status = "okay";
 };
 
-&cls0_trip3 {
+&cls0_trip2 {
 	temperature = <115000>;
 };
 
-&cls1_trip3 {
+&cls1_trip2 {
 	temperature = <115000>;
 };
 
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_deb1.dts b/arch/riscv/boot/dts/spacemit/k1-x_deb1.dts
index b06526852872..05a475d59b86 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_deb1.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_deb1.dts
@@ -953,11 +953,11 @@ &rcpu {
 	status = "okay";
 };
 
-&cls0_trip4 {
+&cls0_trip2 {
 	temperature = <115000>;
 };
 
-&cls1_trip4 {
+&cls1_trip2 {
 	temperature = <115000>;
 };
 
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_deb2.dts b/arch/riscv/boot/dts/spacemit/k1-x_deb2.dts
index bced90483a40..eace7ed52b34 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_deb2.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_deb2.dts
@@ -917,11 +917,11 @@ &rcpu {
 	status = "okay";
 };
 
-&cls0_trip4 {
+&cls0_trip2 {
 	temperature = <115000>;
 };
 
-&cls1_trip4 {
+&cls1_trip2 {
 	temperature = <115000>;
 };
 
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_evb.dts b/arch/riscv/boot/dts/spacemit/k1-x_evb.dts
index 9b1163279110..4a9919dd2d22 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_evb.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_evb.dts
@@ -854,11 +854,11 @@ &rcpu {
 	status = "okay";
 };
 
-&cls0_trip4 {
+&cls0_trip2 {
 	temperature = <115000>;
 };
 
-&cls1_trip4 {
+&cls1_trip2 {
 	temperature = <115000>;
 };
 
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_hs450.dts b/arch/riscv/boot/dts/spacemit/k1-x_hs450.dts
index abff4eca0d9f..92f8dd268bb0 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_hs450.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_hs450.dts
@@ -756,11 +756,11 @@ &rcpu {
 	status = "okay";
 };
 
-&cls0_trip4 {
+&cls0_trip2 {
 	temperature = <115000>;
 };
 
-&cls1_trip4 {
+&cls1_trip2 {
 	temperature = <115000>;
 };
 
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_kx312.dts b/arch/riscv/boot/dts/spacemit/k1-x_kx312.dts
index a00fe73fae9f..5062ea7bafb3 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_kx312.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_kx312.dts
@@ -725,11 +725,11 @@ &rcpu {
 	status = "okay";
 };
 
-&cls0_trip4 {
+&cls0_trip2 {
 	temperature = <115000>;
 };
 
-&cls1_trip4 {
+&cls1_trip2 {
 	temperature = <115000>;
 };
 
diff --git a/arch/riscv/boot/dts/spacemit/k1-x_mingo.dts b/arch/riscv/boot/dts/spacemit/k1-x_mingo.dts
index 7ec431db70c4..19a10eea0232 100644
--- a/arch/riscv/boot/dts/spacemit/k1-x_mingo.dts
+++ b/arch/riscv/boot/dts/spacemit/k1-x_mingo.dts
@@ -602,10 +602,10 @@ &rcpu {
 	status = "okay";
 };
 
-&cls0_trip3 {
+&cls0_trip2 {
 	temperature = <115000>;
 };
 
-&cls1_trip3 {
+&cls1_trip2 {
 	temperature = <115000>;
 };
diff --git a/drivers/cpufreq/spacemit-cpufreq.c b/drivers/cpufreq/spacemit-cpufreq.c
index 6d47a155d9d7..5fd734cffa0a 100644
--- a/drivers/cpufreq/spacemit-cpufreq.c
+++ b/drivers/cpufreq/spacemit-cpufreq.c
@@ -124,13 +124,12 @@ static int spacemit_processor_notifier(struct notifier_block *nb,
 
 	if (event == CPUFREQ_PRECHANGE) {
 
+		mutex_lock(&regulator_mutex);
+
 		if (freqs->new > freqs->old) {
 			/* increase voltage first */
-			if (vol_qos[cpu]->regulator) {
-				mutex_lock(&regulator_mutex);
+			if (vol_qos[cpu]->regulator)
 				freq_qos_update_request(&vol_qos[cpu]->qos, microvol / 1000);
-				mutex_unlock(&regulator_mutex);
-			}
 		}
 
 		/**
@@ -170,12 +169,11 @@ static int spacemit_processor_notifier(struct notifier_block *nb,
 
 		if (freqs->new < freqs->old) {
 			/* decrease the voltage last */
-			if (vol_qos[cpu]->regulator) {
-				mutex_lock(&regulator_mutex);
+			if (vol_qos[cpu]->regulator)
 				freq_qos_update_request(&vol_qos[cpu]->qos, microvol / 1000);
-				mutex_unlock(&regulator_mutex);
-			}
 		}
+
+		mutex_unlock(&regulator_mutex);
 	}
 
 	dev_pm_opp_put_opp_table(opp_table);
-- 
2.47.0

